src/openrct2/actions/SignSetStyleAction.hpp

```diff
@@ -0,0 +1,137 @@
+    SignSetStyleAction() = default;
+    SignSetStyleAction(int32_t bannerIndex, uint8_t mainColour, uint8_t textColour, bool isLarge)
+        : _bannerIndex(bannerIndex)
+        , _mainColour(mainColour)
+        , _textColour(textColour)
+        , _isLarge(isLarge)
+    {
+    }
+
+    uint16_t GetActionFlags() const override
+    {
+        return GameAction::GetActionFlags() | GA_FLAGS::ALLOW_WHILE_PAUSED;
+    }
+
+    void Serialise(DataSerialiser & stream) override
+    {
+        GameAction::Serialise(stream);
+        stream << DS_TAG(_bannerIndex) << DS_TAG(_mainColour) << DS_TAG(_textColour) << DS_TAG(_isLarge);
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        if ((BannerIndex)_bannerIndex >= MAX_BANNERS || _bannerIndex < 0)
+        {
+            log_warning("Invalid game command for setting sign style, banner id '%d' out of range", _bannerIndex);
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        rct_banner* banner = &gBanners[_bannerIndex];
+
+        int32_t x = banner->x << 5;
+        int32_t y = banner->y << 5;
+
+        if (_isLarge)
+        {
+            TileElement* tileElement = banner_get_tile_element((BannerIndex)_bannerIndex);
+            if (tileElement == nullptr)
+            {
+                log_warning("Invalid game command for setting sign style, banner id '%d' not found", _bannerIndex);
+                return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+            }
+            if (tileElement->GetType() != TILE_ELEMENT_TYPE_LARGE_SCENERY)
+            {
+                log_warning("Invalid game command for setting sign style, banner id '%d' is not large", _bannerIndex);
+                return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+            }
+        }
+        else
+        {
+            TileElement* tileElement = map_get_first_element_at(x / 32, y / 32);
+            bool wallFound = false;
+            do
+            {
+                if (tileElement->GetType() != TILE_ELEMENT_TYPE_WALL)
+                    continue;
+
+                rct_scenery_entry* scenery_entry = tileElement->AsWall()->GetEntry();
+                if (scenery_entry->wall.scrolling_mode == 0xFF)
+                    continue;
+                if (tileElement->AsWall()->GetBannerIndex() != (BannerIndex)_bannerIndex)
+                    continue;
+                wallFound = true;
+                break;
+            } while (!(tileElement++)->IsLastForTile());
+
+            if (!wallFound == false)
+            {
+                log_warning("Invalid game command for setting sign style, banner id '%d' not found", _bannerIndex);
+                return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+            }
+        }
+
+        return MakeResult();
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        rct_banner* banner = &gBanners[_bannerIndex];
+
+        int32_t x = banner->x << 5;
```

Comment:
While we're at it, could you save the x and y as `CoordsXY`? Code should look something like
`CoordsXY coords = { banner->x * 32, banner->y * 32 };`

The advantage of Coords/TileCoords is that it actually tells whether it saves big or small x/y coordinates, as well as keeping them together. You should also be able to pass `CoordsXY` into `map_get_first_element_at()` directly.